### PR TITLE
Implement cache import/export (includes auto-generated config files)

### LINKED ISSUE #9: Cache file created for operations that require existing cache
**Description**

Certain operations don't make sense when the application isn't pointed to an existing, valid cache file.

* `cache-control --clear`
* `cache-control --export-file <file>`
* `start --mode forward`

Right now, the cache file is opened in read/write mode, which appears to create a new file at the specified path if it doesn't exist. For the aforementioned operations, this might create confusion for a user that has mistyped the filename and expects the operation to be performed on an existing cache file.

**Steps to reproduce**
* Run any of the aforementioned operations (e.g. `cache-control --clear`) with `cache.cache_file_path` set to a non-existent file.
* A new cache file will exist at the specified path

**Expected behavior**

* If clearing cache, exporting a cache file, or running in forward mode, check if the cache file exists. If it doesn't then report an error.
* If importing cache or running in cache or store mode, create the cache file if it doesn't exist and write a log message that the cache file has been created

### LINKED ISSUE #10: Implement the cache import and export options
**Is your feature request related to a problem? Please describe.**
In order to use the store and forward options effectively, there needs to be a way of exporting the stored requests that have no response, and importing forwarded responses to those requests.

(The import/export options are already defined but marked as not implemented in #7)

**Describe the solution you'd like**
- Implement the `export-file` option by finding all the cached activation and deactivation requests without responses (which are those that are auto-forwarded by `forward` mode of operation), and exporting them to the named separate sqlite database which can be transported to a forwarding proxy.  If the named export database already exists (e.g., from a prior export), then append to it.
- Implement the `import-file` option by importing any activation requests and responses where the responses are newer than (or have different activation keys than) the existing activation requests and responses, and by removing any existing activation requests and responses for which (newer) matching deactivation responses exist in the import database.

**Describe alternatives you've considered**
There are two open questions about store/forward functionality:
1. When handling deactivations imported from a forwarding server, should we remove *all* activation responses from the local server, or just the *older* activation responses.  The only chance of a difference between the two behaviors comes if the server you are importing into is running in `cache` rather than `store` mode, which is arguably an error if you are importing forwarded responses.
2. If there are newer activation requests for the same keys in the local server than there are in the forwarded responses, should we remove those requests in favor of the requests that were used by the forwarding server, or should we just import the responses and use them against future requests?

**Additional context**
If you think about how the cache works, you quickly realize that no client of a store server ever gets the response to its original request.  Instead, the stored response from a forwarded request is returned after being "re-id'd" to match the client request.  Thus, there isn't much point to importing activation requests from the forwarded server.

### LINKED ISSUE #12: move to using sqlx master rather than local fork of it
**Description**
Now that launchbadge/sqlx#983 (support for modern tokio) has been merged to SQLX master, we should change our build to pull from SQLX again rather than our (out-of-date) fork of the source of that PR.

Let's wait a day or two to see if they do the 0.5beta release they are talking about doing, in which case we can just use that tag as the source.  Otherwise we can grab a stable commit that has these changes.

### LINKED ISSUE #13: Auto-create configuration file
**Is your feature request related to a problem? Please describe.**
I want to be able to create a configuration file without knowing how to edit TOML format.  Also, I'd like some explanation of the configuration options in the context of creating the file.  This is especially important given that a configuration file is required to operate the proxy.

**Describe the solution you'd like**
I'd like the proxy, when first started, to prompt me for my configuration settings, with explanations.  I'd also like to be able to reconfigure easily by being re-interviewed.

**Describe alternatives you've considered**
Trying to specify all options as command line arguments is tedious and requires me to edit a script file.  I don't want to edit a file to use the proxy.

**Additional context**
This is the same mechanism used by many existing installation tools for sysadmins.

### LINKED ISSUE #16: Storing requests and responses is not transactional
**Description**
Forcing the server to quit can leave no record of recently completed requests in the database.
If I force a core-dump quickly after completing a request on a proxy in store mode, the request is not captured in the database, so that deleting the sqlite temp files causes the next open of the database not to see any record of the transaction.

**Steps to reproduce**
* Run the proxy in `store` mode with `frl-proxy start --mode store`.
* Try a deactivation request that fails.
* Soon thereafter, force-quit the server (not with Ctrl-C but something that stops it dead).
* Clean up the sqlite process temp files (`...-shm` and `...-wal`).
* Check the database to see the record of the deactivation request.  It won't be there.

**Expected behavior**
You should see the deactivation request in the database.

**Environment**
 - frl-proxy version: 0.9

### PR DESCRIPTION
**Pull request by adobeDan**
## Summary
* Implement import and export of requests/responses for use with an external forwarder
* Rework of cache internals to support use of external database files
* Adds transactional saves of requests and responses to cache (fixes #16)
* Includes commits from PR #14 to auto-create configuration files
* Updates all dependencies to latest minor or patch releases
* Update all submodules to use `eyre` for error reporting

## Testing Steps for auto-creation of config files (#13)
* Invoke the proxy in an empty directory.
* A default configuration file will be auto-created.
* A configuration interview will be done based on the default configuration.
* The default file will be updated with the results of the interview.

## Testing steps for import/export and external forwarding (#10)
* Get a package for Photoshop that uses `https://frl-proxy.brotsky.net:8843` as its proxy address
* Configure your proxy to run at `127.0.0.1:8443` with SSL and add an entry to your hosts file mapping 127.0.01 to `frl-proxy.brotsky.net`.  Use the attached `pkcs` certificate file (password `test`) for your testing.  Use the defaults for proxy database file (`frl-proxy.cache`) and proxy log file (`frl-proxy.log`), and set the log destination to `file`.  Use any logging level you want.
* Install 2 different releases (this year and last year) of Photoshop, and install the FRL proxy license package.
* Run your proxy in `store` mode with `frl-proxy start --mode store`
* Launch both versions of Photoshop.  Both will complain that they cannot verify subscription status.
* Quit your proxy.
* Export the unanswered requests with `frl-proxy export frl-proxy.forward`.  It will report exporting 2 requests.
* Move the `frl-proxy.cache` file to `frl-proxy.stored`
* Move the `frl-proxy.forward` file to `frl-proxy.cache`
* Run your proxy in `forward` mode with `frl-proxy start --mode forward`.  It will automatically stop.
* Move the `frl-proxy.cache` file to `frl-proxy.forward`
* Move the `frl-proxy.stored` file to `frl-proxy.cache`
* Import the forwarded data with `frl-proxy import frl-proxy.forward`.  It will report importing 2 request/response pairs.
* Run your proxy in `store` mode with `frl-proxy start --mode store`.
* Launch both versions of Photoshop.  Both will license.
* Stop your proxy.
* Remove the `frl-proxy.forward` file.
* Run the `adobe-licensing-toolkit -l` to verify your activated licenses.  The output should look something like this, with the `NpdId` coming from the license file of your installation:
```bash
$ adobe-licensing-toolkit -l
Adobe Licensing Toolkit (1.1.0.98)
License Details

NpdId          : ODU0YjU5OGQtOTE1Ni00NDZiLWFlZDYtMGQ1ZGM2ZmVhZDBi (*)
AppId          : Bridge1
DeploymentMode : FRL_ONLINE

NpdId          : ODU0YjU5OGQtOTE1Ni00NDZiLWFlZDYtMGQ1ZGM2ZmVhZDBi (*)
AppId          : Photoshop1
DeploymentMode : FRL_ONLINE
(*) indicates currently active(used) license
Operation Successfully Completed
```
* Run the proxy in `store` mode with `frl-proxy start --mode store`
* Deactivate the Photoshop license with `adobe-licensing-toolkit -t -n your-npdId-from-above`.  It will fail.
* Stop your proxy.
* Export the unanswered requests with `frl-proxy export frl-proxy.forward`.  It will report exporting 1 request.
* Move the `frl-proxy.cache` file to `frl-proxy.stored`
* Move the `frl-proxy.forward` file to `frl-proxy.cache`
* Run your proxy in `forward` mode with `frl-proxy start --mode forward`.  It will automatically stop.
* Move the `frl-proxy.cache` file to `frl-proxy.forward`
* Move the `frl-proxy.stored` file to `frl-proxy.cache`
* Import the forwarded data with `frl-proxy import frl-proxy.forward`.  It will report importing 1 request/response pair.
* Clear the Photoshop license from your credential store.
* Run your proxy in `store` mode with `frl-proxy start --mode store`.
* Launch both versions of Photoshop.  Both will complain that they cannot verify subscription status.
* Stop your proxy.

NOTE: Once you have successfully activated Photoshop, it's not quite enough to deactivate it to repeat the tests.  You also have to remove the cached license from the credential store.  The license will have a key that is a base 64 encoded form of the app ID and the certificate generation it uses.  You can use the [adobe-license-decoder](https://github.com/adobe/adobe-license-decoder.rs) utility to show you what that base64 ID is, because it's the first part of the OperatingConfig name of the license file.

## Related Issues and PRs
Fixes #9 
Fixes #10 
Fixes #12
Fixes #13
Fixes #16
Supersedes #14 
Supersedes #15

### CODE REVIEW

**Comment by adobeDan** [comment]
@adorton-adobe This PR supersedes the other two I had opened waiting for review.  It includes all of their changes, so I am going to close them.  Just focus the review on this one, which hopefully takes us all the way to v1.0.

**Comment by adorton-adobe** [comment]
Sounds good. I'll take a look. Once we get this merged, I will do a 1.0.0rc1 release while I work on updating the docs.

**Comment by adorton-adobe** [comment]
The config init workflow breaks when I use the shift key (e.g. when typing `:` for the host port). Seems to be related to this:

https://github.com/mitsuhiko/console/issues/83

Not sure what we can do until that crate gets an update. I'll be sure to mention this issue in the docs. It should be fine since `0.0.0.0:8443` is a reasonable default and the user can edit the config directly after generating it.

**Comment by adobeDan** [comment]
I am also happy to update the host parsing to understand any delimiter between the host and port (e.g., space, hyphen, or other non-shifted delimiters :).  We could also specify host and port in separate config entries, which actually would allow to (for example) specify one port for use with SSL and another for use with non-SSL which would help make the `--ssl` mode flag more useful.

**Comment by adorton-adobe** [comment]
Sounds good

**Comment by adorton-adobe** [comment]
This is also a factor in other options (file paths, remote host), but I don't think that will be too much of an issue.

**Comment by adobeDan** [comment]
Here is the cert file needed for testing: [frl-proxy.pkcs.zip](https://github.com/adobe/frl-online-proxy/files/5992802/frl-proxy.pkcs.zip).

**Comment by adobeDan** [comment]
I notice that mitsuhiko/console#83 has a comment that says this problem is due to the `windows-console-colors` feature.  Have you tried explicitly loading the crate with that feature disabled?  Maybe that would trump the default features used by dialoguer?

**Comment by adorton-adobe** [comment]
After I `import` the forward file, the `-shm` and `-wal` files are still present in my working directory, which I believe might mean we aren't closing the database connection when importing. I'll check the code.

```
-a----         2/17/2021   5:22 PM          32768 frl-proxy.forward-shm
-a----         2/17/2021   5:22 PM              0 frl-proxy.forward-wal
```

**Comment by adobeDan** [comment]
I just checked the code and we definitely close the database we're importing from.  But I am noticing the same behavior.  Here is the [reference article from SQLite](https://sqlite.org/tempfiles.html) about those files.  I will check to see if there is anything in the sqlx community about this.

**Review by adorton-adobe — CHANGES_REQUESTED** [review]
Looks great - I just have a few suggested tweaks.

One other thing to note is when testing cached deactivation, import/export reported importing/exporting 4 requests. 3 of those are my prior activations (Photoshop, Bridge, Acrobat). Should those have been cleared from the cache when I activated those licenses? The rest of the cached deactivation workflow worked as expected - those licenses are inactive.

**Comment by adobeDan** [comment]
Terrific, thanks @adorton-adobe.   I understand and am great with all your suggestions and will implement them today or tomorrow.

WRT your question about the number of request/response pairs: what you saw is expected and correct.  When operating in forward mode, the proxy replays all the stored requests/responses in order and stores the results without processing them.  Then, during the import of the forwarding database, the results are "replayed" as if they had just come in to a live proxy.  So first the two activation requests are processed and stored (waiting for the clients to come back and find them) and then the deactivation response is processed: it is this processing which removes the already received activation responses (and doesn't store the deactivation response because those are never replayed to clients).  So once the import has finished, any clients who come back and request an activation response will not find one, so the client side matches the server side.

File: .changelog/changelog.md
Comment (adobeDan) [reply .changelog/changelog.md:27]: reverted to what's in dev already.

File: src/cli.rs
Comment (adobeDan) [reply src/cli.rs:14]: removed version number

File: .changelog/latest.md
Comment (adobeDan) [reply .changelog/latest.md:1]: Removed.

File: src/settings.rs
Comment (adobeDan) [reply src/settings.rs:92]: Changed all three places it appears.

**Comment by adobeDan** [comment]
@adorton-adobe On the older PR you mentioned that you wanted to change the configuration file suffix back to `toml`.  Do you want me to make that change here?  Do we want to dismiss with the `frl-` prefix and just go with a naming convention like `proxy-conf.toml`, `proxy-cache.sqlite`, `proxy-cert.pfx` and so on?  (I want to change the cert file suffix to `pfx` anyway because that's the 3-letter convention for `pkcs12` format.). That way the file types are transparent.

**Comment by adorton-adobe** [comment]
I think it's totally fine to keep the `frl` prefix, since the proxy is designed to work with FRL Online packages. I suggested the switch back to `toml` for technical reasons - it'll help ensure proper syntax highlighting. We can change it here or in another PR/commit, doesn't really matter to me.

**Comment by adobeDan** [comment]
RIght, I understood why you wanted to change it to toml, and I guess I'm saying I think that really applies to all of the filenames: I think it's way better to use file extensions to reflect the content type of the file and embed the function of the file in the name.  Also, the reason to take out the `frl-` prefix is to separate the files from the executable itself, so that it's safer to delete files without accidentally deleting the executable.  I've been working with this setup:
```
frl-proxy[.exe] (executable)
proxy-cache.sqlite
proxy-cert.pfx
proxy-conf.toml
proxy-log.log
```
and it looks great.  I'd like to check it in, if you're OK with that.

**Comment by adorton-adobe** [comment]
Looks good. Once you push those changes, I think we can merge.

**Comment by adobeDan** [comment]
OK, changes pushed.  Please approve and merge once it finishes building.

**Review by adorton-adobe — APPROVED** [review]

**Pull request merged**
